src/System.Console/tests/SyncTextReader.cs

```diff
@@ -0,0 +1,206 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using Xunit;
+
+public class SyncTextReaderTests
+{
+    // NOTE: These tests test the underlying SyncTextReader by
+    // accessing the Console.In TextReader (which in fact is a SyncTextReader).
+
+    static readonly string[] s_testLines = new string[] {
+        "3232 Hello32 Hello 5032 Hello 50 532 Hello 50 5 aTrueaabcdbc1.23123.4561.23439505050System.ObjectHello World",
+        "32",
+        "",
+        "32 Hello",
+        "",
+        "32 Hello 50",
+        "",
+        "32 Hello 50 5",
+        "",
+        "32 Hello 50 5 a",
+        "",
+        "True",
+        "a",
+        "abcd",
+        "bc",
+        "1.23",
+        "123.456",
+        "1.234",
+        "39",
+        "50",
+        "50",
+        "50",
+        "System.Object",
+        "Hello World",
+    };
+
+    private static void Test(string content, Action<StreamReader> action)
+    {
+        TextWriter savedStandardOutput = Console.Out;
+        TextReader savedStandardInput = Console.In;
+
+        try
+        {
+            using (MemoryStream memStream = new MemoryStream())
+            {
+                using (StreamWriter sw = new StreamWriter(memStream))
+                {
+                    sw.Write(content);
+                    sw.Flush();
+
+                    memStream.Seek(0, SeekOrigin.Begin);
+
+                    using (StreamReader sr = new StreamReader(memStream))
+                    {
+                        Console.SetIn(sr);
+                        action(sr);
+                    }
+                }
+            }
+        }
+        finally
+        {
+            Console.SetOut(savedStandardOutput);
+            Console.SetIn(savedStandardInput);
+        }
+    }
+
+    [Fact]
+    public void ReadToEnd()
+    {
+        var expected = string.Join(Environment.NewLine, s_testLines);
+        Test(expected, reader => {
+            // Given, When
+            var result = Console.In.ReadToEnd();
+
+            // Then
+            Assert.Equal(expected, result);
+            Assert.Equal(-1, Console.Read()); // We should be at EOF now.
+        });
+    }
+
+    [Fact]
+    public void ReadBlock()
+    {
+        var expected = new[] { 'H', 'e', 'l', 'l', 'o' };
+        Test(new string(expected), reader => {
+            // Given
+            var buffer = new char[5];
+
+            // When
+            var result = Console.In.ReadBlock(buffer, 0, 5);
+
+            // Then
+            Assert.Equal(5, result);
+            Assert.Equal(expected, buffer);
+            Assert.Equal(-1, Console.Read()); // We should be at EOF now. 
+        });
+    }
+
+    [Fact]
+    public void Read()
+    {
+        var expected = new[] { 'H', 'e', 'l', 'l', 'o' };
+        Test(new string(expected), reader => {
+            // Given
+            var buffer = new char[5];
```

Comment:
Rather than using a hardcoded 5 in all of the locations that care about the length, consider just using expected.Length.  That way if you change the contents of expected, everything else changes with it.